CICD.Tools.CatalogUpload.Lib/CatalogMetaDataFactory.cs

```diff
@@ -0,0 +1,325 @@
+				}
+			}
+
+			description = ($"Minimum DataMiner Version: {minimumDmaVersion}\r\n{description}");
+
+			if (!String.IsNullOrWhiteSpace(buildNumber))
+			{
+				meta.ArtifactHadBuildNumber = true;
+				// Throw away the CU version. If we have a build number it's a pre-release.
+				string version = appInfo.Element("Version")?.Value;
+
+				if (version != null)
+				{
+					if (version.Contains("-CU"))
+					{
+						// Throw away the CU version. If we have a build number it's a pre-release.
+						version = version.Split('-')[0];
+					}
+
+					meta.Version.Value = version + "-B" + buildNumber;
+					description = $"Pre-Release (Unofficial) version.\r\n{description}";
+				}
+			}
+			else
+			{
+				meta.ArtifactHadBuildNumber = false;
+				meta.Version.Value = appInfo.Element("Version")?.Value;
+				description = $"{description ?? "No Description."}";
+			}
+
+			if (description.Length > 500) description = description.Substring(0, 497) + "...";
+
+
+			meta.Version.VersionDescription = description;
+			meta.ContentType = contentType;
+			return meta;
+		}
+
+		private static CatalogMetaData FromDmprotocol(string pathToDmprotocol)
+		{
+			// Description.txt
+			/*
+				Protocol Name: Microsoft Platform
+				Protocol Version: 6.0.0.4_B2
+			 * */
+
+			string descriptionFileDMProtocolText;
+			string protocolXmlString;
+
+			using (var zipFile = ZipFile.OpenRead(pathToDmprotocol))
+			{
+				var foundFile = zipFile.Entries.FirstOrDefault(x => x.Name.Equals("Description.txt", StringComparison.InvariantCulture));
+				if (foundFile == null) throw new InvalidOperationException("Could not find Description.txt in the .dmprotocol.");
+
+				using (var stream = foundFile.Open())
+				{
+					using (var memoryStream = new StreamReader(stream))
+					{
+						descriptionFileDMProtocolText = memoryStream.ReadToEnd();
+					}
+				}
+
+				var foundProtocol = zipFile.Entries.FirstOrDefault(x => x.Name.EndsWith("Protocol.xml", StringComparison.InvariantCulture));
+				if (foundProtocol == null) throw new InvalidOperationException("Could not find Protocol.xml in the .dmprotocol.");
+
+				using (var stream = foundProtocol.Open())
+				{
+					using (var memoryStream = new StreamReader(stream))
+					{
+						protocolXmlString = memoryStream.ReadToEnd();
+					}
+				}
+			}
+
+			CatalogMetaData meta = new CatalogMetaData();
+
+			var lines = descriptionFileDMProtocolText.Split(new[] { "\r\n", "\r", "\n" }, StringSplitOptions.None);
+
+			foreach (var line in lines)
+			{
+				var splitLine = line.Split(':');
+
+				switch (splitLine[0])
+				{
+					case "Protocol Name":
+						meta.Name = splitLine[1].Trim();
+						break;
+
+					case "Protocol Version":
+						meta.Version.Value = splitLine[1].Trim();
+						break;
+
+					default:
+						break;
+				}
+			}
+
+			// Note. Tried using Skyline.DataMiner.CICD.Parsers but could not find a method to get VersionHistory.
+			// So... doing this again... sigh.
+
```

Comment:
What was the problem?

Reply:
I could not find a method to retrieve the VersionHistory as a text in the library. Was taking more time than it takes to parse it from the XDocument. So eventually wrote the parsing myself. I've removed the unprofessional comment.